experimental/parser/parse_type.go

```diff
@@ -0,0 +1,196 @@
+// will emit diagnostics to that effect. As such, the current token position on
+// c should not be nil.
+//
+// May return nil if parsing completely fails.
+func parseType(p *parser, c *token.Cursor, where taxa.Place, pathAfter bool) (ast.TypeAny, ast.Path) {
+	var isList, isInMethod bool
+	switch where.Subject {
+	case taxa.MethodIns, taxa.MethodOuts,
+		taxa.KeywordReturns: // Used when parsing the invalid `returns foo.Bar` production.
+		isInMethod = true
+		fallthrough
+	case taxa.TypeParams:
+		isList = true
+	}
+
+	// First, parse a path, possibly preceded by a sequence of modifiers.
+	//
+	// To do this, we repeatedly parse paths, and each time we get a path that
+	// starts with an identifier, we interpret it as a modifier. For example,
+	// repeated.foo needs to be interpreted as repeated .foo.
+	var (
+		mods   []token.Token
+		tyPath ast.Path
+	)
+	for !c.Done() && tyPath.Nil() {
+		next := c.Peek()
+		if !canStartPath(next) {
+			break
+		}
+
+		tyPath = parsePath(p, c)
+		if tyPath.Absolute() {
+			break // Absolute paths cannot start with a modifier, so we are done.
+		}
+
+		first, _ := iters.First(tyPath.Components)
+		ident := first.AsIdent()
+		if ident.Nil() {
+			break // If this starts with an extension, we're done.
+		}
+
+		// Here is a nasty case. Suppose the user has written within message
+		// scope something like
+		//
+		//  package .foo.bar = 5;
+		//
+		// This is a syntax error but only because the name of the field is a
+		// non-trivial path. However, we would like for this to be diagnosed as
+		// a package declaration. Thus, if this looks like a bad package, we
+		// break it up so that the type is "package" and the path is ".foo.bar",
+		// so that the DeclPackage code path can diagnose it.
+		//
+		// We require that this have no modifiers and that it not be followed by
+		// a path, so that the following productions are *not* treated as weird
+		// packages:
+		//
+		//  optional package.foo.bar;
+		//  package.foo.bar baz;
+		//
+		// Note that this does not apply inside of type lists. This is because
+		// type lists *only* contain types, and not productions started by
+		// keywords.
+		//
+		// This case applies to many other keywords. See cannotStartType.
+		if !isList && len(mods) == 0 &&
+			cannotStartType(ident.Text()) &&
+			!canStartPath(c.Peek()) {
+			kw, path := tyPath.Split(1)
+			if !path.Nil() {
+				return ast.TypePath{Path: kw}.AsAny(), path
+			}
+		}
+
+		// Check if ident is a modifier, and if so, peel it off.
+		//
+		// We need to be careful to only peel off `stream` inside of a method
+		// type. If the entire path is a single identifier, we always peel it
+		// off, since code that follows handles turning it back into a path
+		// based on what comes after it.
+		var isMod bool
+		_, rest := tyPath.Split(1)
+		switch ast.TypePrefixByName(ident.Name()) {
+		case ast.TypePrefixOptional, ast.TypePrefixRepeated, ast.TypePrefixRequired:
+			isMod = !isList || rest.Nil()
```

Comment:
If `isInMethod` is true, these prefixes are not actually allowed. I guess it will be up to the legalizer to handle this and possibly reject it?

It may also need to fuse a prefix with its following type since `stream optional.foo.bar` is allowed, but should be interpreted as having a single prefix `stream`, *not* as two prefixes `stream` and `optional` with a type name of `.foo.bar`.

Reply:
I added a comment explaining why the parse you suggest won't happen.